// @dronesense/core/lib/common/enums/Autopilot.js
"use strict";
(function (Autopilot) {
    Autopilot[Autopilot["Unknown"] = 0] = "Unknown";
    Autopilot[Autopilot["Generic"] = 1] = "Generic";
    Autopilot[Autopilot["Reserved"] = 2] = "Reserved";
    Autopilot[Autopilot["Slugs"] = 3] = "Slugs";
    Autopilot[Autopilot["ArdupilotMega"] = 4] = "ArdupilotMega";
    Autopilot[Autopilot["OpenPilot"] = 5] = "OpenPilot";
    Autopilot[Autopilot["GenericWaypointsOnly"] = 6] = "GenericWaypointsOnly";
    Autopilot[Autopilot["GenericWaypointsAndSimpleNavigationOnly"] = 7] = "GenericWaypointsAndSimpleNavigationOnly";
    Autopilot[Autopilot["GenericMissionFull"] = 8] = "GenericMissionFull";
    Autopilot[Autopilot["Invalid"] = 9] = "Invalid";
    Autopilot[Autopilot["PPZ"] = 10] = "PPZ";
    Autopilot[Autopilot["UAVDevBoard"] = 11] = "UAVDevBoard";
    Autopilot[Autopilot["FlexiPilot"] = 12] = "FlexiPilot";
    Autopilot[Autopilot["PX4"] = 13] = "PX4";
    Autopilot[Autopilot["SMACCMPilot"] = 14] = "SMACCMPilot";
    Autopilot[Autopilot["AutoQuad"] = 15] = "AutoQuad";
    Autopilot[Autopilot["Armazila"] = 16] = "Armazila";
    Autopilot[Autopilot["Aerob"] = 17] = "Aerob";
    Autopilot[Autopilot["ASLUAV"] = 18] = "ASLUAV";
})(exports.Autopilot || (exports.Autopilot = {}));
var Autopilot = exports.Autopilot;
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = Autopilot;
