// @dronesense/model/src/models/OrbitCommand.ts
// import Command from '../models/Command';
// import GeoPoint from '../models/GeoPoint';
// import jquery from 'jquery';

// export default class OrbitCommand extends Command {

//     // Center Point of POI
//     Position: GeoPoint;

//     // In meters, above ground
//     Altitude: number;

//     // In meters per second
//     Speed: number;

//     // Clockwise or Counter-Clockwise
//     RotationDirection: any;

//     // Radius of the circle in meters
//     RadiusDistance: number;

//     // Angle of camera between 0-90 in degrees
//     CameraAngle: number;

//     // Degrees of rotation between photo captures
//     PhotoInterval: number;

//     HeadingMode: any;

//     HeadingValue: number;

//     PhotoIntervalMode: any;

//     PhotoIntervalValue: number;

//     private _data: any = {};

//     constructor(id?: string, handle?: any) {
//         super('Orbit', id, handle);

//         if (this.handle) {
//             this.hydrate(this.handle.data);
//             this.handle.on(this.handle.EVENTS.DATA_CHANGED, this.hydrate.bind(this));
//         }

//     }

//     hydrate(data: any): void {

//         if (jQuery.isEmptyObject(data)) {
//             this.handle.off(this.handle.EVENTS.DATA_CHANGED);
//             return;
//         }

//         if (data) {

//             var dataChanged: boolean = false;

//             if (!this._data.Position ||
//                 this._data.Position.lat !== data.Position.lat ||
//                 this._data.Position.lng !== data.Position.lng) {
//                 this.Position = new GeoPoint(data.Position.lat, data.Position.lng);
//                 dataChanged = true;
//             }

//             if (this._data.Altitude !== data.Altitude) {
//                 this.Altitude = data.Altitude;
//                 dataChanged = true;
//             }

//             if (this._data.Speed !== data.Speed) {
//                 this.Speed = data.Speed;
//                 dataChanged = true;
//             }

//             if (this._data.RotationDirection !== data.RotationDirection) {
//                 this.RotationDirection = data.RotationDirection;
//                 dataChanged = true;
//             }

//             if (this._data.RadiusDistance !== data.RadiusDistance) {
//                 this.RadiusDistance = data.RadiusDistance;
//                 dataChanged = true;
//             }

//             if (this._data.CameraAngle !== data.CameraAngle) {
//                 this.CameraAngle = data.CameraAngle;
//                 dataChanged = true;
//             }

//             if (this._data.PhotoInterval !== data.PhotoInterval) {
//                 this.PhotoInterval = data.PhotoInterval;
//                 dataChanged = true;
//             }

//             if (this._data.HeadingMode !== data.HeadingMode) {
//                 this.HeadingMode = data.HeadingMode;
//                 dataChanged = true;
//             }

//             if (this._data.HeadingValue !== data.HeadingValue) {
//                 this.HeadingValue = data.HeadingValue;
//                 dataChanged = true;
//             }

//             if (this._data.PhotoIntervalMode !== data.PhotoIntervalMode) {
//                 this.PhotoIntervalMode = data.PhotoIntervalMode;
//                 dataChanged = true;
//             }

//             if (this._data.PhotoIntervalValue !== data.PhotoIntervalValue) {
//                 this.PhotoIntervalValue = data.PhotoIntervalValue;
//                 dataChanged = true;
//             }

//             if (this._data.Order !== data.Order) {
//                 this.Order = data.Order;
//                 dataChanged = true;
//             }

//             if (this._data.Name !== data.Name) {
//                 this.Name = data.Name;
//                 dataChanged = true;
//             }

//             if (this._data.Expanded !== data.Expanded) {
//                 this.Expanded = data.Expanded;
//                 dataChanged = true;
//             }

//             this._data = data;

//             if (dataChanged) {
//                 //this.onModelUpdated.trigger('dataChanged');
//             }

//         }
//     }

//     SaveProperty(prop: any, propName: string): void {
//         this.handle.setValue(propName, prop)
//             .catch((error: any): void => {
//                 // TODO how should we display error?
//                 console.log(error);
//             });
//     }

//     Update(data: any): void {
//         this.handle.update(data).catch((error: any): void => {
//             console.log(error);
//         });
//     }

//     static defaultData(position: GeoPoint, radius: number): any {
//         return {
//             Position: position,
//             Altitude: 60,
//             Speed: 5,
//             RotationDirection: 'Clockwise',
//             RadiusDistance: radius,
//             CameraAngle: 45,
//             PhotoInterval: 4,
//             HeadingMode: 'Toward Center',
//             HeadingValue: 0,
//             PhotoIntervalMode: 'Degrees',
//             PhotoIntervalValue: 10,
//         };
//     }
// }
