// @dronesense/model/lib/common/Enums.js
"use strict";
(function (Unit) {
    Unit[Unit["Standard"] = 0] = "Standard";
    Unit[Unit["Metric"] = 1] = "Metric";
})(exports.Unit || (exports.Unit = {}));
var Unit = exports.Unit;
(function (CoordinatePreference) {
    CoordinatePreference[CoordinatePreference["Decimal"] = 0] = "Decimal";
    CoordinatePreference[CoordinatePreference["Degrees"] = 1] = "Degrees";
})(exports.CoordinatePreference || (exports.CoordinatePreference = {}));
var CoordinatePreference = exports.CoordinatePreference;
(function (PositionType) {
    PositionType[PositionType["North"] = 0] = "North";
    PositionType[PositionType["RelativeToSubjectCourse"] = 1] = "RelativeToSubjectCourse";
})(exports.PositionType || (exports.PositionType = {}));
var PositionType = exports.PositionType;
(function (CircleDirection) {
    CircleDirection[CircleDirection['Clockwise'] = 0] = 'Clockwise';
    CircleDirection[CircleDirection['CounterClockwise'] = 1] = 'CounterClockwise';
})(exports.CircleDirection || (exports.CircleDirection = {}));
var CircleDirection = exports.CircleDirection;
(function (FrameType) {
    FrameType[FrameType["Quadcopter"] = 0] = "Quadcopter";
    FrameType[FrameType["Hexacopter"] = 1] = "Hexacopter";
    FrameType[FrameType["Octocopter"] = 2] = "Octocopter";
    FrameType[FrameType["Airplane"] = 3] = "Airplane";
})(exports.FrameType || (exports.FrameType = {}));
var FrameType = exports.FrameType;
(function (BatteryType) {
    BatteryType[BatteryType["LiPo3S"] = 0] = "LiPo3S";
    BatteryType[BatteryType["LiPo4S"] = 1] = "LiPo4S";
    BatteryType[BatteryType["LiPo5S"] = 2] = "LiPo5S";
    BatteryType[BatteryType["LiPo6S"] = 3] = "LiPo6S";
    BatteryType[BatteryType["LiPo7S"] = 4] = "LiPo7S";
    BatteryType[BatteryType["LiPo8S"] = 5] = "LiPo8S";
})(exports.BatteryType || (exports.BatteryType = {}));
var BatteryType = exports.BatteryType;
(function (AutopilotType) {
    AutopilotType[AutopilotType['APM'] = 0] = 'APM';
    AutopilotType[AutopilotType['3DR Pixhawk'] = 1] = '3DR Pixhawk';
    AutopilotType[AutopilotType['DJI'] = 2] = 'DJI';
})(exports.AutopilotType || (exports.AutopilotType = {}));
var AutopilotType = exports.AutopilotType;
(function (HeadingMode) {
    HeadingMode[HeadingMode['Follow Flight Path'] = 0] = 'Follow Flight Path';
    HeadingMode[HeadingMode['Fixed Heading'] = 1] = 'Fixed Heading';
})(exports.HeadingMode || (exports.HeadingMode = {}));
var HeadingMode = exports.HeadingMode;
(function (AltGSDMode) {
    AltGSDMode[AltGSDMode['Altitude'] = 0] = 'Altitude';
    AltGSDMode[AltGSDMode['GSD'] = 1] = 'GSD';
})(exports.AltGSDMode || (exports.AltGSDMode = {}));
var AltGSDMode = exports.AltGSDMode;
(function (POIHeadingModes) {
    POIHeadingModes[POIHeadingModes['Toward Center'] = 0] = 'Toward Center';
    POIHeadingModes[POIHeadingModes['Along Flight Path'] = 1] = 'Along Flight Path';
    POIHeadingModes[POIHeadingModes['Fixed Heading'] = 2] = 'Fixed Heading';
    POIHeadingModes[POIHeadingModes['Away from Center'] = 3] = 'Away from Center';
})(exports.POIHeadingModes || (exports.POIHeadingModes = {}));
var POIHeadingModes = exports.POIHeadingModes;
(function (PhotoIntervalMode) {
    PhotoIntervalMode[PhotoIntervalMode['Degrees'] = 0] = 'Degrees';
    PhotoIntervalMode[PhotoIntervalMode['Distance'] = 1] = 'Distance';
    PhotoIntervalMode[PhotoIntervalMode['Time'] = 2] = 'Time';
})(exports.PhotoIntervalMode || (exports.PhotoIntervalMode = {}));
var PhotoIntervalMode = exports.PhotoIntervalMode;
var EnumEx = (function () {
    function EnumEx() {
    }
    EnumEx.getNames = function (e) {
        var a = [];
        for (var val in e) {
            if (isNaN(parseInt(val, 10))) {
                a.push(val);
            }
        }
        return a;
    };
    EnumEx.getValues = function (e) {
        var a = [];
        for (var val in e) {
            if (!isNaN(parseInt(val, 10))) {
                a.push(parseInt(val, 10));
            }
        }
        return a;
    };
    return EnumEx;
}());
exports.EnumEx = EnumEx;
