// @dronesense/core/src/common/entities/IPosition.ts
// MavLink: GLOBAL_POSITION_INT
// https://pixhawk.ethz.ch/mavlink/#GLOBAL_POSITION_INT

/**
 * DJI Mobile SDK Values (Android)
 * DJIFlightController
 *  getCurrentState(): DJIFlightControllerDataType.DJIFlightControllerCurrentState
 *
 * DJIFlightControllerDataType.DJIFlightControllerCurrentState
 *  getAircraftLocation(): DJIFlightControllerDataType.DJILocationCoordinate3D
 *  getGpsSignalStatus(): DJIFlightControllerDataType.DJIGPSSignalStatus
 *  getVelocityX(): float // meters/sec
 *  getVelocityY(): float // meters/sec
 *  getVelocityZ(): float // meters/sec
 *  getSatelliteCount(): double
 *
 * DJIFlightControllerDataType.DJILocationCoordinate3D
 *  getAltitude() : float
 *  getCoordinate2D() : DJIFlightControllerDataType.DJILocationCoordinate2D
 *  getLatitude(): double
 *  getLongitude(): double
 *
 * DJIFlightControllerDataType.DJILocationCoordinate2D
 *  getLatitude(): double
 *  getLongitude(): double
 *
 * DJIFlightControllerDataType.DJIGPSSignalStatus (enum)
 *  Level0 - very bad signal
 *  Level1
 *  Level2
 *  Level3
 *  Level4
 *  Level5 - very good signal
 *  None
 */

// GPS_RAW_INT, GLOBAL_POSITION_INT
// export default class GlobalPosition {
// 	// GLOBAL_POSITION_INT
// 	lattitude: number;
// 	longitude: number;
// 	altitude: number;
// 	relative_altitude: number;  // not in DJI?
// 	vx: number;
// 	vy: number;
// 	vz: number;
// 	heading: number;

// 	// GPS_RAW_INT
// 	fix_type: FixType;          // available in NazaCanDecoder, not in DJI sdk?
// 	// latitude: number;
// 	// longitude: number;
// 	// altitude: number;
// 	hdop: number;               // in NazaCanDecoder, not in DJI SDK? - use DJIGPSSignalStatus
// 	vdop: number;               // in NazaCanDecoder, not in DJI SDK?
// 	velocity: number;           // not in DJI?
// 	courseOverGround: number;   // in NazaCanDecoder, not in DJI SDK?
// 	satellites_visible: number;
// 	raw: any;
// }

import { FixType } from '../enums/FixType';

export interface IPosition {
    timestamp: number;
    fixType: FixType;
    hdop: number;
    vdop: number;
    courseOverGround: number;
    heading: number;
    satelliteCount: number;
    groundSpeed: number;
    lattitude: number;
    longitude: number;
    altitudeMSL: number;
    altitudeAGL: number;
    xVelocity: number;
    yVelocity: number;
    zVelocity: number;
}

export default IPosition;
