// @dronesense/core/src/common/entities/DJI/IDJIBattery.ts
export interface IDJIBattery {
    // The total amount of energy, in mAh (milliamp hours), stored in the battery when the battery is fully charged.
    fullChargeEnergy: number;

    // The remaining energy stored in the battery in mAh (milliamp hours).
    currentEnergy: number;

    // The current battery voltage (mV).
    currentVoltage: number;

    // The real time current draw of the battery (mA). A negative value means the battery is being discharged, and a positive value means it is being charged.
    currentCurrent: number;

    // The percentage of remaining lifetime value of the battery. The range of this value is [0 - 100].
    lifetimeRemainingPercent: number;

    // The percentage of battery energy left. The range of this value is [0 - 100].
    batteryEnergyRemainingPercent: number;

    // The temperature of battery in Centigrade, with a range [-128 to 127].
    batteryTemperature: number;

    // The total number of discharges the battery has gone through over its lifetime. The total number of discharges includes discharges that happen through normal use and discharges that are manually set.
    numberOfDischarge: number;

    // Current cell voltage level of the battery. It is only supported when the connected product is stand-alone A3.
    cellVoltageLevel: number;

    // YES if the battery is being charged. It is only supported by Osmo Mobile.
    isBeingCharged: boolean;
}

export default IDJIBattery;
