// @dronesense/model/lib/models/commands/SensorCaptureCommand.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var WaypointCommand_1 = require('./WaypointCommand');
var SensorCaptureCommand = (function (_super) {
    __extends(SensorCaptureCommand, _super);
    function SensorCaptureCommand(type, id, handle, isBase) {
        _super.call(this, type, id, handle, true);
        this.isBase = isBase;
        if (this.handle) {
            this.hydrate(this.handle.data);
            if (!isBase) {
                this.handle.on(this.handle.EVENTS.DATA_CHANGED, this.hydrate.bind(this));
            }
        }
    }
    SensorCaptureCommand.prototype.hydrate = function (data) {
        if (_.isEmpty(data)) {
            this.handle.off(this.handle.EVENTS.DATA_CHANGED);
            return;
        }
        if (data) {
            var dataChanged = false;
            dataChanged = _super.prototype.hydrate.call(this, data);
            if (this.Heading !== data.Heading) {
                this.Heading = data.Heading;
                dataChanged = true;
            }
            if (this.Pitch !== data.Pitch) {
                this.Pitch = data.Pitch;
                dataChanged = true;
            }
            if (this.Roll !== data.Roll) {
                this.Roll = data.Roll;
                dataChanged = true;
            }
            if (dataChanged && !this.isBase) {
            }
            return dataChanged;
        }
    };
    return SensorCaptureCommand;
}(WaypointCommand_1.WaypointCommand));
exports.SensorCaptureCommand = SensorCaptureCommand;
